### ISSUE TITLE
When DNS flap, haproxy ends with 503 on backends

### ISSUE THREAD

**Issue by skacfjin** [haproxy/haproxy#1286]
## Detailed description of the problem
When DNS flap (or VPN tunnel behind which DNS is), then HAProxy ends with `Jun 11 11:56:38 localhost haproxy[3586]: Server backend_node1 is going DOWN for maintenance (DNS timeout status). 1 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.` (all backends go down) and HAProxy ends with responding 503 to all requests. DNS/VPN flap is usually in about 2 minutes.

I have to restart HAProxy to get it working again.

## Expected behavior

HAProxy should get DNS connection back after DNS is working again

## Steps to reproduce the behavior

1. Get configuration below
2. Stop both DNS / break the connection towards both DNS for 2 minutes

## Do you have any idea what may have caused this?
No :(

## Do you have an idea how to solve the issue?
No :(

## What is your configuration?

```
global
    log         127.0.0.1 local2

    chroot      /var/lib/haproxy
    pidfile     /var/run/haproxy.pid
    maxconn     4000
    user        haproxy
    group       haproxy
    daemon

    stats socket 127.0.0.1:14567

    tune.ssl.default-dh-param  2048
    ssl-default-bind-ciphers   ECDH+AESGCM:ECDH+CHACHA20:ECDH+AES256:ECDH+AES128:!aNULL:!SHA1
    ssl-default-bind-options   force-tlsv12 prefer-client-ciphers # force tls1.2 only
    ssl-default-server-ciphers ECDH+AESGCM:ECDH+CHACHA20:ECDH+AES256:ECDH+AES128:!aNULL:!SHA1
    ssl-default-server-options force-tlsv12 # tlsv1.2 only!
    ssl-server-verify required

#---------------------------------------------------------------------
# common defaults that all the 'listen' and 'backend' sections will
# use if not designated in their block
#---------------------------------------------------------------------
defaults
    mode                    http
    log                     global
    option                  httplog
    option                  dontlognull
    option http-server-close
    option forwardfor       except 127.0.0.0/8
    option                  redispatch
    option  log-health-checks
    retries                 3
    timeout http-request    10s
    timeout queue           1m
    timeout connect         10s
    timeout client          1m
    timeout server          1m
    timeout http-keep-alive 10s
    timeout check           10s
    maxconn                 3000
    default-server init-addr last,libc,none resolvers dnsservers

    errorfile 503 /etc/haproxy/errors/503.http

resolvers dnsservers
    nameserver servername <ip>:53
    nameserver server2name <ip2>:53
    resolve_retries 300
    timeout retry 2s
    hold valid 60s
    hold timeout 3600s
    hold refused 3600s
    hold obsolete 3600s
    hold other 3600s

frontend main
    bind *:443 ssl crt <certificate_path>

    # log host header
    http-request capture req.hdr(Host) len 50

    # HSTS
    http-response set-header Strict-Transport-Security max-age=16000000

    use_backend servers

backend servers
    balance roundrobin
    cookie ULTIMATECOOKIE insert indirect nocache

    server server_node1 <servername>:<port> ssl ca-file <ca_certificate_path> check cookie s1 resolvers dnsservers
    server server_node2 <servername>:<port> ssl ca-file <ca_certificate_path> check cookie s2 resolvers dnsservers
```

## Output of `haproxy -vv` and `uname -a`

```
[root@server ~]# haproxy -vv
HA-Proxy version 1.8.26 2020/08/03
Copyright 2000-2020 Willy Tarreau <willy@haproxy.org>

Build options :
  TARGET  = linux2628
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label
  OPTIONS = USE_LINUX_TPROXY=1 USE_CRYPT_H=1 USE_GETADDRINFO=1 USE_ZLIB=1 USE_REGPARM=1 USE_OPENSSL=1 USE_LUA=1 USE_SYSTEMD=1 USE_PCRE=1

Default settings :
  maxconn = 2000, bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with OpenSSL version : OpenSSL 1.0.2k-fips  26 Jan 2017
Running on OpenSSL version : OpenSSL 1.0.2k-fips  26 Jan 2017
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : SSLv3 TLSv1.0 TLSv1.1 TLSv1.2
Built with Lua version : Lua 5.3.4
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Encrypted password support via crypt(3): yes
Built with multi-threading support.
Built with PCRE version : 8.32 2012-11-30
Running on PCRE version : 8.32 2012-11-30
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Built with zlib version : 1.2.7
Running on zlib version : 1.2.7
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with network namespace support.

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available filters :
        [SPOE] spoe
        [COMP] compression
        [TRACE] trace

[root@server ~]# uname -a
Linux <hostname> 3.10.0-1160.15.2.el7.x86_64 #1 SMP Thu Jan 21 16:15:07 EST 2021 x86_64 x86_64 x86_64 GNU/Linux
```

## Additional information (if helpful)
I tried to change hold values, but now we get this error after one hour after flap (aka 3600s).

**Comment by lukastribus** [comment]
1.8 is a quasi-unsupported branch (critical fixes only), and 1.8.26 is not even the latest version of it (lacking 5 - 6 different DNS fixes compared to 1.8.30 and probably a ton of DNS fixes compared to supported release branches).

I strongly suggest you upgrade to a supported LTS release branch (see haproxy.org).

**Comment by skacfjin** [comment]
We are using UIS repo, because they provide redhat packages, since haproxy.org is providing only tar.gz :)

I checked the repo and they are providing latest 1.8.28-1.el7.ius, but I will try to migrate to haproxy22.x86_64 and report back :)

**Comment by git001** [comment]
Well https://github.com/DBezemer/rpm-haproxy/releases provides also rpm packages even the latest one :-)

**Comment by capflam** [comment]
I guess this bug was fixed in 2.3 by this [commit](http://git.haproxy.org/?p=haproxy-2.3.git;a=commit;h=751872e4). It was backported to 2.2. And I'm on 2.0 backports. The 2.4 is not affected by this bug.

As @lukastribus said, it is probably a good idea to move on 2.0 or 2.2. However, for now, there is no release shipping this fix.

**Comment by skacfjin** [comment]
Thanks! I will try to plan migration to 2.4 and lets see :-) Closing for now, if there will be anything I will reopen :)

**Comment by skacfjin** [comment]
Hello people of haproxy! :)

So we did upgrade (finally! :D) to version 2.2.19-1.el7.ius.x86_64, but the issue with DNS is still there. When we tried manually flap VPN, then after 10 minutes (resolve_retries [300] * timeout retry [2s]). 

From changelog, I see the issue which @capflam meantion is fixed in:
```
2021/07/16 : 2.2.15
    ...
    - BUG/MEDIUM: dns: reset file descriptor if send returns an error
    ...
```
What we saw from tcpdump is, that haproxy have still one outgoing port from which it's creating request to DNS.. any idea if it's possible to change it to create totally new connection?

**Comment by wtarreau** [comment]
It seems from the patch that it's what is happening upon error.

**Comment by wtarreau** [comment]
Hello,

just wondering, after your VPN reconnection, are you seeing DNS packets leave the machine ? If so, are you seeing responses come ? If not, could you figure why ? (e.g. bad address somewhere etc). Because what I suspect is that after the VPN is down, haproxy sees errors on send() and closes the connection, then tries again later. If you don't see any packet it would indicate that for some reason there's a faulty state internally. If you're seeing these packets and no response, then the problem is on the path to the DNS server.

**Comment by skacfjin** [comment]
Hi,

so what we saw in tcpdump is following:

haproxy is hitting dns from same source ip (and port) when VPN flap and it's keeping doing it after vpn is up again (still same port and ip), so looks like haproxy not closing session, but using still the same session (which is broken, since vpn flapped) instead of using new session

We had also while true; do nslookup google.com; sleep 10; done and that was working well - we were able to hit dns, but it's creating always new session

This is the configuration right now which we have for dns resolvers:
```
resolvers our-dns
    nameserver dns1 10.10.10.20:53
    nameserver dns2 10.10.10.10:53
    resolve_retries 300
    timeout retry 2s
    hold valid 60s
    hold timeout 120s
    hold refused 120s
    hold obsolete 120s
    hold other 120s
```

**Comment by wtarreau** [comment]
Thanks. I suspect that for whatever reason no error is reported by the OS when the VPN is down and that it might be why the connection does not change. Probably that an improvement would consist in having a new parameter that says after how long a silent nameserveer ought to be considered unreachable so that a new connection could be initiated.

Just thinking loud, that might be a good use case for DNS over TCP instead of UDP. Apparently in 2.4 you already have TCP-baesd DNS, that can be enabled by specifying `tcp@` in front of the address. I don't know how the timeouts are set, but that makes me think that it would be convenient to enable `tcp-ut` there that is already supported on regular servers.

CCing @EmericBr in case he has ideas on all this.

**Comment by skacfjin** [comment]
But we have right now version 2.2 :-/. We are also investigating idea that firewall is keeping the connection (there is default timeout for connection) in some "limbo" and that is the reason why haproxy doenst create new connection. So if firewall/vpn would close that connection, haproxy might create a new one.

**Comment by wtarreau** [comment]
The issue you're facing is that there's no such real thing as a connection in UDP, and what your firewall, VPN and others are doing around it is doing the best they can. UDP was never designed to be used over intermittent links with a dynamic address so if at some point the local association with the destination is lost due to a disappearing interface, and the route switches to the default route for example, your packets go into the wild with no error. Same if the interface comes back but the IP changed. Unless you're able to enforce rules on the firewall to say that non-local addresses must be rejected, there's no way any process on the machine is aware that this happened. This is first and foremost an architecture problem. We can try to workaround it to some extents but it will never be perfect due to the inherent nature of an unreliable protocol sent over a disappearing medium that doesn't cause errors to flow back :-/

**Comment by lukastribus** [comment]
> But we have right now version 2.2 :-/. We are also investigating idea that firewall is keeping the connection (there is default timeout for connection) in some "limbo" and that is the reason why haproxy doenst create new connection. So if firewall/vpn would close that connection, haproxy might create a new one.

Yes I suggest you take a look at UDP session timeouts on your Firewall. UDP session timeouts for port 53 can be very small. I would suggest (for UDP port 53) 10 seconds, certainly not larger then 30 seconds.

You can also check whether clearing the session on the firewall in this situation restores connectivity (just to confirm this theory).

**Comment by wtarreau** [comment]
Lukas, I'm not sure the UDP session will be recreated on inactivity, I think it only happens when an I/O error is reported, and the problem with UDP is that your packets can be sent in the ether and ignored without the application code ever noticing. A similar case that can cause such trouble is when binding to an interface that disappears and reappears. It's not the same anymore once it reappers, so packets follow a different route...

**Comment by lukastribus** [comment]
Yes, I'm pretty sure that haproxy doesn't destroy and recreate the socket due to a DNS response timeout. So we probably keep hitting the DNS server from the same source port, matching the same session in a stateful layer4 firewall (or NAT device). If this session table in this stateful layer4 device points to a broken destination interface (for example due to an interface flap combined with a bug that is: existing sessions are not properly updated on interface/VPN flaps), then those DNS packets will go down the ether.

That's why I'm saying if we can reduce the session timeout *on that stateful layer4 device* to match a single DNS "transaction", we will probably solve the issue (at it's core).

I'm not sure recreating sockets in haproxy would be a good idea just to workaround this (external) problem.

**Comment by wtarreau** [comment]
Given that the problem could happen on the local machine itself, I *tend* to think it could be useful. But based on the existing code it's probably not something trivial to do.

**Comment by maximepoulin-hl** [comment]
We've started having a very similar issue around 3-4 weeks ago, but without any VPN involved. Our HAproxies all have public IP addresses directly connected to the Internet with no NAT. For some reason DNS seems to have a brief blip, and our backends go down and never come back up. We use 1.1.1.1 as our DNS is hosted by Cloudflare. We do use DNS over UDP, haven't tried TCP to see if it fixes it. Also haven't tried sending it to systemd-resolved on 127.0.0.53 but that would probably also make it more reliable (but we wouldn't get as much of an authoritative DNS answer).

```
[WARNING] Server $backend/app1 is going DOWN for maintenance (DNS timeout status). 2 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
[WARNING] Server $backend/app2 is going DOWN for maintenance (DNS timeout status). 1 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
[WARNING] Server $backend/app3 is going DOWN for maintenance (DNS timeout status). 0 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
[ALERT] backend $backend has no server available!
```

We were using Ubuntu's version in the repos for Focal (20.04), version `2.0.29-0ubuntu1.3`. We've upgraded to 2.6.x with the PPA to version `2.6.8-2ppa1~focal` in hope it might be fixed as we saw a fair amount of work was done on DNS, but the issue persists.

We think it may be related to some network issues on our OpenStack provider's end, but seems to exhibit a bug in HAproxy as a result.

I've tried reproducing by randomly dropping DNS with iptables, but every time it recovers just fine. However, when we have our network blips, it reliably never recovers, it's permanently stuck unless manually reloaded or some automation triggers a reload to update another backend or frontend. Unfortunately this happens maybe once or twice a week on one of 8 servers, so I don't have much information. All I can say is whatever causes HAproxy to get stuck in an invalid state seems to be a result of external network issues. Possibly duplicate DNS answers, or getting the answer after it timed out?

**Comment by kakawait** [comment]
> We've started having a very similar issue around 3-4 weeks ago, but without any VPN involved. Our HAproxies all have public IP addresses directly connected to the Internet with no NAT. For some reason DNS seems to have a brief blip, and our backends go down and never come back up. We use 1.1.1.1 as our DNS is hosted by Cloudflare. We do use DNS over UDP, haven't tried TCP to see if it fixes it. Also haven't tried sending it to systemd-resolved on 127.0.0.53 but that would probably also make it more reliable (but we wouldn't get as much of an authoritative DNS answer).
> 
> ```
> [WARNING] Server $backend/app1 is going DOWN for maintenance (DNS timeout status). 2 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
> [WARNING] Server $backend/app2 is going DOWN for maintenance (DNS timeout status). 1 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
> [WARNING] Server $backend/app3 is going DOWN for maintenance (DNS timeout status). 0 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
> [ALERT] backend $backend has no server available!
> ```
> 
> We were using Ubuntu's version in the repos for Focal (20.04), version `2.0.29-0ubuntu1.3`. We've upgraded to 2.6.x with the PPA to version `2.6.8-2ppa1~focal` in hope it might be fixed as we saw a fair amount of work was done on DNS, but the issue persists.
> 
> We think it may be related to some network issues on our OpenStack provider's end, but seems to exhibit a bug in HAproxy as a result.
> 
> I've tried reproducing by randomly dropping DNS with iptables, but every time it recovers just fine. However, when we have our network blips, it reliably never recovers, it's permanently stuck unless manually reloaded or some automation triggers a reload to update another backend or frontend. Unfortunately this happens maybe once or twice a week on one of 8 servers, so I don't have much information. All I can say is whatever causes HAproxy to get stuck in an invalid state seems to be a result of external network issues. Possibly duplicate DNS answers, or getting the answer after it timed out?

Did you find some solutions?

**Comment by JooKop** [comment]
We are experiencing the exact same issue, commenting for visibility in case someone knows a workaround for this problem.

We encountered this issue in a Kubernetes environment using the HAProxy directive `parse-resolv-conf`. Now we have added additional nameservers for `1.1.1.1` and `8.8.8.8` into the mix and hoping it will solve at least a part of network related issues.

**Comment by maximepoulin-hl** [comment]
We still have this issue as well, happened again last month. We've switched it to TCP mode in the hope that the TCP stack will somehow make sure it doesn't deadlock. I'm waiting to see if it happens again and my next step will be to route everything through a local DNS proxy such that DNS never flaps.

If that doesn't solve it I'll just template it directly in the config and call it a day. DNS was always a stopgap solution for us anyway.

Version is `2.6.18-1ppa1~focal 2024/06/20`.

**Comment by capflam** [comment]
A gentle ping to check if we can do something for the 3.1 or not.

**Comment by lucascicco** [comment]
We’re encountering a similar issue with Kubernetes' Kube-DNS. Our setup uses server-template to dynamically generate backends, leveraging a resolver whose nameserver is the Kube-DNS service IP. However, the Kubernetes service IP is intercepted by kube-proxy, ultimately resolving to the pod IP.

The problem arises when one of the Kube-DNS replicas briefly goes down. During this downtime, HAProxy logs a DNS timeout error and occasionally gets stuck, requiring a manual restart to recover.

Is there a fix for it? or workaround?

Please keep me updated.
